reflect/src/main/java/dagger/reflect/Scope.java

```diff
@@ -252,7 +252,7 @@ Builder addBindingElementsIntoSet(Key key, Binding elementsBinding) {
       if (key == null) throw new NullPointerException("key == null");
       if (elementsBinding == null) throw new NullPointerException("elementsBinding == null");
       if (Types.getRawType(key.type()) != Set.class) {
-        throw new IllegalArgumentException("key.type() must be Set");
+        throw new IllegalArgumentException("@BindsIntoSet must return Set");
```

Comment:
I have tentative plans to expose `Scope` as public API. It shouldn't know anything about Dagger and should function like a general-purpose JSR 330-ish service locator. So I would prefer to keep this exception message the same (or, well, it should really print the offending key...).

Reply:
Ah, I'll update it to print the key. So duplicate the exception in both places?
I wanted to add another check here in a follow up PR similar to the check in Dagger which says `@Binds methods must return a primitive, an array, a type variable, or a declared type` if the return type is `Set<? extends String?`.

Should we move the check out to a common place? Or will your upcoming changes to expose Scope as a public API move these checks somewhere else?

Reply:
Yeah duplicate it. Think about it like this: Scope is validating what it knows about which is only keys mapped to bindings and it wants to prevent a duplicate. `@BindsIntoSet` is validating a method shape for that specific annotation in order to create a key and binding to insert into a scope.

I'll try to pull out the non-Dagger-specific bits tomorrow or next week. Any naming suggestions? It's completely Dagger-free so it can be anything.

Reply:
What about Sheath? It's where you stick your daggers!

Reply:
That used to be a thing for getting Dagger running on GWT: https://github.com/tbroyer/sheath. Seems like it could be used.